Xamarin.Forms.Core/Performance.cs

```diff
@@ -1,95 +1,34 @@
-﻿using System;
-using System.Collections.Generic;
-using System.ComponentModel;
-using System.Diagnostics;
-using System.Linq;
+﻿using System.ComponentModel;
 using System.Runtime.CompilerServices;
-using System.Text;
 
 namespace Xamarin.Forms.Internals
 {
 	[EditorBrowsable(EditorBrowsableState.Never)]
-	public static class Performance
+	public interface IPerformanceProvider
 	{
-		static readonly Dictionary<string, Stats> Statistics = new Dictionary<string, Stats>();
+		void Stop(string reference, string tag, string path, string member);
 
-		[Conditional("PERF")]
-		public static void Clear()
-		{
-			Statistics.Clear();
-		}
-
-		public static void Count(string tag = null, [CallerFilePath] string path = null, [CallerMemberName] string member = null)
-		{
-			string id = path + ":" + member + (tag != null ? "-" + tag : string.Empty);
-
-			Stats stats;
-			if (!Statistics.TryGetValue(id, out stats))
-				Statistics[id] = stats = new Stats();
-
-			stats.CallCount++;
-		}
-
-		[Conditional("PERF")]
-		public static void DumpStats()
-		{
-			Debug.WriteLine(GetStats());
-		}
-
-		public static string GetStats()
-		{
-			var b = new StringBuilder();
-			b.AppendLine("ID                                                                                 | Call Count | Total Time | Avg Time");
-			foreach (KeyValuePair<string, Stats> kvp in Statistics.OrderBy(kvp => kvp.Key))
-			{
-				string key = ShortenPath(kvp.Key);
-				double total = TimeSpan.FromTicks(kvp.Value.TotalTime).TotalMilliseconds;
-				double avg = total / kvp.Value.CallCount;
-				b.AppendFormat("{0,-80} | {1,-10} | {2,-10}ms | {3,-8}ms", key, kvp.Value.CallCount, total, avg);
-				b.AppendLine();
-			}
-			return b.ToString();
-		}
-
-		[Conditional("PERF")]
-		public static void Start(string tag = null, [CallerFilePath] string path = null, [CallerMemberName] string member = null)
-		{
-			string id = path + ":" + member + (tag != null ? "-" + tag : string.Empty);
-
-			Stats stats;
-			if (!Statistics.TryGetValue(id, out stats))
-				Statistics[id] = stats = new Stats();
+		void Start(string reference, string tag, string path, string member);
+	}
 
-			stats.CallCount++;
-			stats.StartTimes.Push(Stopwatch.GetTimestamp());
-		}
+	[EditorBrowsable(EditorBrowsableState.Never)]
+	public class Performance
+	{
+		public static IPerformanceProvider Provider { get; private set; }
 
-		[Conditional("PERF")]
```

Comment:
why are you removing the `Conditional` ?

Reply:
I have replaced the `Conditional` with a null object pattern. If no provider is found, then the code is no-op. The `Conditional` can't be there if we want to allow our users to build their own performance trackers, which was part of my requirements. Core code is compiled by the time it makes it to users.